src/OrchardCore/OrchardCore.Environment.Shell.Abstractions/Builders/ShellContext.cs

```diff
@@ -58,75 +61,121 @@ public IServiceScope EnterServiceScope()
         /// </summary>
         public int ActiveRequests => _refCount;
 
-        public void RequestStarted()
+        /// <summary>
+        /// Returns whether the shell can be used for new requests.
+        /// </summary>
+        public bool RequestStarted()
         {
+            if (_released)
+            {
+                return false;
+            }
+
             Interlocked.Increment(ref _refCount);
+
+            return true;
         }
 
-        public void RequestEnded()
+        /// <summary>
+        /// Returns whether the shell can be released as a result of the request being ended.
+        /// </summary>
+        public bool RequestEnded()
         {
             var refCount = Interlocked.Decrement(ref _refCount);
+            return _released && refCount == 0;
         }
 
-        public bool CanTerminate => _released && _refCount == 0;
+        public bool CanTerminate => _released && _refCount == 0 && !_disposed;
 
         /// <summary>
         /// Mark the <see cref="ShellContext"/> has a candidate to be released.
         /// </summary>
         public void Release()
         {
+            if (_released == true)
+            {
+                // Prevent infinite loops with circular dependencies
+                return;
+            }
+
             // When a tenant is changed and should be restarted, its shell context is replaced with a new one, 
             // so that new request can't use it anymore. However some existing request might still be running and try to 
             // resolve or use its services. We then call this method to count the remaining references and dispose it 
             // when the number reached zero.
 
             _released = true;
-        }
-
-        public void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
 
-        protected virtual void Dispose(bool disposing)
-        {
-            if (!_disposed)
+            lock (this)
```

Comment:
Removing the double check lock?

Reply:
We can't because the other method locks for creating the reference.